File: c/stack/stack.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include "stack.h"

void initStack(Stack *stack){
    stack->top = -1;
}

bool isEmpty(Stack* stack){
    return stack->top == MAX_SIZE - 1;
}

bool isFull(Stack* stack){
    return MAX_SIZE - 1 == stack->top;
}

void push(Stack* stack, int data){
    if (isFull(stack)){
        fprintf(stderr, "Stack is full\n");
        return;
    }
    stack->arr[++stack->top] = data;
}

int pop(Stack* stack){
    if (isEmpty(stack)){
        fprintf(stderr, "Stack is empty\n");
        return -1;
    }
    return stack->arr[stack->top--];
}

int peek(Stack* stack){
    if (isEmpty(stack)){
        fprintf(stderr, "Stack is empty\n");
        return -1;
    }
    return stack->arr[stack->top];
}

void printStack(Stack* stack){
    if (isEmpty(stack)){
        fprintf(stdout, "Stack is empty\n");
        return;
    }
    for (int i = 0; i <= stack->top; i++){
        fprintf(stdout, "%d ", stack->arr[i]);
    }
    fprintf(stdout, "\n");
}

File: go/stack/stack.go
package main

import "fmt"

type Stack struct {
	items []int
}

func NewStack() *Stack {
	return &Stack{
		items: make([]int, 0),
	}
}

func (s *Stack) Push(value int) {
	s.items = append(s.items, value)
}

func (s *Stack) Pop() (int, bool) {
	if s.IsEmpty() {
		return 0, false
	}
	n := len(s.items)
	value := s.items[n-1]
	s.items = s.items[:n-1]
	return value, true
}

func (s *Stack) Top() (int, bool) {
	if s.IsEmpty() {
		return 0, false
	}
	return s.items[len(s.items)-1], true
}

func (s *Stack) IsEmpty() bool {
	return len(s.items) == 0
}

func (s *Stack) String() string {
	if s.IsEmpty() {
		return ""
	}
	result := ""
	for i := len(s.items) - 1; i >= 0; i-- {
		result += fmt.Sprintf("%d", s.items[i])
		if i > 0 {
			result += " "
		}
	}
	return result
}
